### PR TITLE
WIP: Improve primitives for gw and vm

### PR DESCRIPTION
**Pull request by grimpy**
Signed-off-by: Jo De Boeck <deboeck.jo@gmail.com>

### CODE REVIEW

**Comment by grimpy** [comment]
This is only meant as specification it's not a full implementation (yet)

**Comment by grimpy** [comment]
Example how to use these promitives:
```
node = j.clients.zeroos.sal.node_get('myzos')
gw = node.gateways.create(
nics=[{'type': 'vlan',
  'id': '0',
  'name': 'public',
  'config': {'cidr': '192.168.58.22/24', 'gateway': '192.168.58.254'}},
 {'type': 'vxlan',
  'id': '100',
  'name': 'private',
  'dhcpserver': {'nameservers': ['8.8.8.8'],
   'hosts': [{'macaddress': '00:A0:C9:14:C8:29',
       'hostname': 'myzeroos',
       'ipaddress': '192.168.112.11',},
     {'macaddress': '00:A0:C9:14:C8:30',
       'hostname': 'myubuntu',
       'ipaddress': '192.168.112.12',},
  ]
  'config': {'cidr': '192.168.112.22/24'}}]
])
zos = node.hypervisor.create_zero_os('myzos', 'https://bootstrap.gig.tech/ipxe/master/abcef01234567890/organization=myorg', nics=[{'type': 'vxlan', 'id': 100, 'hwaddr': '00:A0:C9:14:C8:29'}])
ubu = node.hypervisor.create_zero_os('myubuntu', nics=[{'type': 'vxlan', 'id': 100, 'hwaddr': '00:A0:C9:14:C8:30'}])
```
```

File: JumpScale9Lib/clients/zero_os/sal/Hypervisor.py [hunk 1]
```diff
@@ -28,3 +28,29 @@ def list(self):
 
     def get(self, uuid):
         return VM(uuid, self.node)
+
+    def start_zero_os(self, name, ipxeurl='https://bootstrap.gig.tech/ipxe/master/0', **kwargs):
+        '''
```
Comment (serboctor) [inline JumpScale9Lib/clients/zero_os/sal/Hypervisor.py]: docstrings should use double quotes for sake of consistency and for sake of being pep8 compliant.

File: JumpScale9Lib/clients/zero_os/sal/gateway/__init__.py [hunk 2]
```diff
@@ -1,10 +1,13 @@
 from .gateway import Gateway
 
 class Gateways:
-    def get(self, container, data):
+    def __init__(self, node):
+        self.node = node
+
+    def get(self, name):
         """
         param data: dict object containing detailed information about the gateway
```
Comment (serboctor) [inline JumpScale9Lib/clients/zero_os/sal/gateway/__init__.py]: obsolete param?

File: JumpScale9Lib/clients/zero_os/sal/primitives.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    def create_gateway(self, name):
+        """
+        Create gateway object
+
+        To deploy gatewy invoke .deploy method
+
+        :param name: Name of the gateway
+        :type name: str
+        :type flist: str
```
Comment (serboctor) [inline JumpScale9Lib/clients/zero_os/sal/primitives.py]: obsolete line?

**Pull request merged**